ping-pong/interactor/src/interact.rs

```diff
@@ -1,370 +1,324 @@
     }
 
-    pub async fn upgrade(
-        &mut self,
-        ping_amount: RustBigUint,
-        duration_in_seconds: u64,
-        opt_activation_timestamp: Option<u64>,
-        max_funds: OptionalValue<RustBigUint>,
-    ) {
-        let response = self
+    pub async fn upgrade(&mut self, ping_amount: RustBigUint, duration_in_seconds: u64) {
+        let upgrade_address = self
             .interactor
             .tx()
-            .to(self.state.current_ping_pong_egld_address())
-            .from(&self.wallet_address)
+            .from(&self.alice_wallet_address)
+            .to(self.state.current_ping_pong_address())
             .gas(30_000_000u64)
-            .typed(proxy_ping_pong_egld::PingPongProxy)
-            .upgrade(
-                ping_amount,
-                duration_in_seconds,
-                opt_activation_timestamp,
-                max_funds,
-            )
+            .typed(ping_pong_proxy::PingPongProxy)
+            .upgrade(ping_amount, duration_in_seconds)
             .code(PING_PONG_CODE)
             .returns(ReturnsNewAddress)
             .run()
             .await;
 
-        println!("Result: {response:?}");
+        let upgrade_address_bech32 = bech32::encode(&upgrade_address);
+        self.state
+            .set_ping_pong_address(Bech32Address::from_bech32_string(
+                upgrade_address_bech32.clone(),
+            ));
+
+        println!("new upgrade address: {upgrade_address_bech32}");
     }
 
-    pub async fn ping(&mut self, egld_amount: u64, message: Option<&str>, sender: &Bech32Address) {
-        let response = self
-            .interactor
-            .tx()
-            .from(sender)
-            .to(self.state.current_ping_pong_egld_address())
-            .gas(30_000_000u64)
-            .typed(proxy_ping_pong_egld::PingPongProxy)
-            .ping()
-            .egld(egld_amount)
-            .returns(ReturnsHandledOrError::new())
-            .run()
-            .await;
+    pub async fn ping(
+        &mut self,
+        token_id: &str,
+        nonce: Option<u64>,
+        amount: u64,
+        sender: &Bech32Address,
+        message: Option<&str>,
+    ) {
+        let response = if token_id.to_ascii_uppercase() == "EGLD" {
+            self.interactor
+                .tx()
+                .from(sender)
+                .to(self.state.current_ping_pong_address())
+                .gas(30_000_000u64)
+                .typed(ping_pong_proxy::PingPongProxy)
+                .ping()
+                .egld(amount)
+                .returns(ReturnsHandledOrError::new())
+                .run()
+                .await
+        } else {
+            self.interactor
+                .tx()
+                .from(sender)
+                .to(self.state.current_ping_pong_address())
+                .gas(30_000_000u64)
+                .typed(ping_pong_proxy::PingPongProxy)
+                .ping()
+                .payment((
```

Comment:
We have an EgldOrEsdtTokenPayment that works both ways, to avoid code duplication